keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1150 @@
+}
+
+  // SelectableField specifies the JSON path of a field that may be used with field selectors.
+type SelectableField struct {
+	// jsonPath is a simple JSON path which is evaluated against each custom resource to produce a
+  // field selector value.
+  // Only JSON paths without the array notation are allowed.
+  // Must point to a valid field of type string, boolean or integer. Strings with enum values
+  // and strings with formats are allowed.
+  // Must not point to metdata fields.
+	JSONPath string `json:"jsonPath" protobuf:"bytes,1,opt,name=jsonPath"`
+
+  // valueIfAbsent provides a value for field selection if the value referred to
+  // by jsonPath is absent.
+  // valueIfAbsent is required unless all parts of jsonPath are required fields, or have defaults.
+  // +Optional
+  ValueIfAbsent *string `json:"valueIfAbsent" protobuf:"bytes,2,opt,name=valueIfAbsent"`
+}
+```
+
+##### Validaiton rules
+
+- `selectableFields[].jsonPath` must be a "simple path" (similar to
+  `additionalPrinterColumns[].jsonPath` except validated properly). I.e.
+    property names separated by '.', e.g. `.spec.x.y` (The path may not contain
+    map or array indexing)
+- The schema type referenced by `selectableFields[].jsonPath` must be one of:
+  string, integer, boolean (enums and fields with string formats are allowed)
+- If (and only if) the schema type referenced by `selectableFields[].jsonPath`
+  is optional (directly or indirectly), a `selectableFields[].ifAbsent` field
+  must provide a fallback string value.
+- A maximum of 8 `selectableFields` per CRD version are allowed.
+- `selectableFields` may not refer to metadata fields.
+- `selectableFields` may not contain duplicates.
+
+
+#### Discovery
+
+We do not publish any field selector information into OpenAPIv3, and we do not
+document in APIs or on the Kubernetes website which fields may be selected.
+
+Finding the field selectors for built-in types, for the above table, required
+locating [`GetAttrs/*ToSelectableFields` functions in the Kubernetes source code](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/strategy.go#L340).
+
+We will add discovery details in to OpenAPIv3:
+
+```json
+"io.k8s.api.core.v1.Pod": {
+        "description": "Pod is a collection of containers that can run on a host. This resource is created by clients and scheduled onto hosts.",
+        "properties": { ... },
+        "type": "object",
+        "x-kubernetes-group-version-kind": [ ... ],
+        "x-kubernetes-selectable-fields": [
+          { "fieldPath": "spec.nodeName"},
+          { "fieldPath": "spec.restartPolicy"},
+          { "fieldPath": "spec.schedulerName"},
+          { "fieldPath": "spec.serviceAccountName"},
+          { "fieldPath": "spec.hostNetwork", "valueIfAbsent": "false", "description": "Set to the value of the spec.securityContext.hostNetwork field"},
```

Comment:
We would generate the description based on the field level documentation?

Reply:
We'd produce this only for legacy built-ins that have behavior that additional documentation.

CRDs won't need this because the fieldPath and valueIfAbsent (I'm going to rename this to "default") describe the selectableField sufficiently, so I don't intent to allow CRDs to provide a description.